GH-1784 eval stats

hmottestad opened the pull request:
GitHub issue resolved: #1784  <!-- add a Github issue number here, e.g #123. This line 
                              automatically closes the issue when the PR is merged --> 

Briefly describe the changes proposed in this PR:

* three stats implementations: constant, direct, dynamic
   * constant reuses existing implementation
   * direct queries the underlying data structure
   * dynamic uses arrays of HyperLogLog counters to keep estimates

hmottestad commented:
http://dev.eclipse.org/ipzilla/show_bug.cgi?id=21359

hmottestad commented:
Missing:
 - async & parallel dynamic stats
 - support remove operations from dynamic cache

hmottestad commented:
@jeenbroekstra what do you think of my approach here? I’m using HyperLogLog to count the number of statements for each pattern. 

I have arrays, single and two dimensional. I use parts of the pattern as a key. Eg. For the “subjectsIndex” array statement.getSubject().hashCode() % subjectsIndex.length gives the index into the array. I hash the statement with murmur and store the hash in the hyperloglog counter. 

The two dimensional arrays can index both a subject and predicate, for instance. 

I have two sets of arrays. One for added statements and one for removed. 

The biggest drawback is that it poorly handles removing a statement and the re-adding it. 

This is all still WIP.

jeenbroekstra commented on core/sail/extensible-store/src/main/java/org/eclipse/rdf4j/sail/extensiblestore/evaluationstatistics/ExtensibleDynamicEvaluationStatistics.java:
```diff
@@ -0,0 +1,337 @@
+package org.eclipse.rdf4j.sail.extensiblestore.evaluationstatistics;
```
Copyright header is missing here, and a couple of other places as well.

jeenbroekstra commented on core/sail/extensible-store/src/main/java/org/eclipse/rdf4j/sail/extensiblestore/evaluationstatistics/ExtensibleDynamicEvaluationStatistics.java:
```diff
@@ -0,0 +1,337 @@
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Stream;
+
+public class ExtensibleDynamicEvaluationStatistics extends ExtensibleEvaluationStatistics implements DynamicStatistics {
```
I believe this class, and some of the other implementation classes as well, deserve some extensive javadoc that goes into the how and why of the algorithm used.

jeenbroekstra commented on core/sail/extensible-store/src/main/java/org/eclipse/rdf4j/sail/extensiblestore/evaluationstatistics/ExtensibleDynamicEvaluationStatistics.java:
```diff
@@ -0,0 +1,337 @@
+				.estimateCardinality()
+				- index_removed[Math.abs(value1.hashCode() % index_removed.length)][Math
+						.abs(value2.hashCode() % index_removed.length)]
+								.estimateCardinality();
```
I'll have to believe you if you say that it does what it should do. Seriously though: I wonder if you could maybe split this out into more than one line of code to make it easier to read.

hmottestad replied:
Will do :D - it got a lot worse when I added in the index_removed support.

jeenbroekstra commented on core/sail/extensible-store/src/main/java/org/eclipse/rdf4j/sail/extensiblestore/evaluationstatistics/ExtensibleDynamicEvaluationStatistics.java:
```diff
@@ -0,0 +1,337 @@
+
+		int size = queueSize.incrementAndGet();
+		if (size > QUEUE_LIMIT && queueThread == null) {
+			startQueueThread();
```
What is the role of this thread? It updates the various indexes I gather, is that such an expensive operation that it should be handled asynchronously?

hmottestad replied:
It's to do async stats updates. The thread is started if the size of the queue is likely to be above the limit, then runs until the queue is empty. Since the stats are only estimates there is no guarantee for timeliness. Those guarantees are needed when running tests and benchmarking, so there is a poor mans latch that waits for the thread to finish.

hmottestad commented:
> It looks like a solid enough approach - though I can't pretend to fully understand it. Got some inline editorial comments, including some questions.
> 
> On a more general note: I am curious why you are going with HyperLogLog, and an implementation that, if I'm not mistaken, comes from yet _another_ third-party database (Apache Druid). ElasticSearch itself has an API for retrieving index statistics - wouldn't it be simpler to base your evaluation statistics on that, directly?

I might end up writing something specific for the elasticsearch implementation, but I first wanted to make a generic and scaleable implementation that can be reused for other sails built on the ExtensibleStore. Might not end up making anything specific for the elasticsearch implementation since the network overhead might not be worth it. 

As for this particular HyperLogLog implementation, I'm not married to it. I could shop around a bit more, especially if the CQ drags out. HyperLogLog in general seems like a decent fit though, so all I need is a good-enough implementation.

hmottestad commented on core/sail/elasticsearch-store/src/test/java/org/eclipse/rdf4j/sail/elasticsearchstore/benchmark/DeleteBenchmark.java:
```diff
@@ -60,7 +61,7 @@
 
 		embeddedElastic = TestHelpers.startElasticsearch(installLocation,
-				"/Library/Java/JavaVirtualMachines/jdk1.8.0_144.jdk/Contents/Home");
+				"/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home");
```
Yeah. Because of how benchmark forks the jvm, the JAVA_HOME env isn't being passed through as expected. This was a dirty hack to get it to function.

hmottestad replied:
I'm not going to try to fix this.

jeenbroekstra replied:
Can you add a comment to explain why this is there?

hmottestad replied:
Will do.

jeenbroekstra commented:
Can this be postponed to release 3.2.0? I want to start creating the IP log for 3.1, and if we submit it for review without the Druid CQ, we can't merge this in until after the release.

hmottestad commented:
> Can this be postponed to release 3.2.0? I want to start creating the IP log for 3.1, and if we submit it for review without the Druid CQ, we can't merge this in until after the release.

Sure. This PR isn't finished anyway and might not be done for another few weeks.

hmottestad commented:
![B38A58CA-68B4-42AB-A189-963D5A3DA6A9](https://user-images.githubusercontent.com/797185/73052674-ee6e4d00-3e85-11ea-8ff4-e19ed2760248.png)

That there doesn’t look particularly compatible with our license does it?

jeenbroekstra commented:
From the pic it looks like it's Apache licence, that's fine to use for us. As is BSD licence.

hmottestad commented on core/sail/base/src/main/java/org/eclipse/rdf4j/sail/base/SailSourceConnection.java:
```diff
@@ -657,17 +656,17 @@ public void clearInferred(Resource... contexts) throws SailException {
+				remove(null, null, null, inferredOnlyDataset, inferredOnlySink, contexts);
 			}
-			inferredSink.clear(contexts);
+			inferredOnlySink.clear(contexts);
```
@jeenbroekstra 

I found out that the SailSourceConnection will get all statements and remove them from the inferred sink, instead of getting just the inferred statements. So I fixed that by introducing a tri-state enum, essentially you can now specify: all, inferredOnly or explicitOnly. This is only for this class....and `branch(...)` is a private method, so no breaking changes. 

I don't have any decent benchmarks for the rdfs sail that add and remove statements, but I do have some for the SPIN sail and that's seeing a 20% speed bump for some of the benchmarks.

jeenbroekstra replied:
I'm lost. Is this related to #1795 ?

hmottestad replied:
Not related no. This just makes clearing inferred statements faster.

hmottestad commented:
@jeenbroekstra I'm getting ready to merge this now. Want to start testing it through work. I've cleaned up the things you pointed out, added some documentation to the classes, cleaned up and added some more tests. 

The biggest thing missing from this is support for storing and loading evaluation statistics between restarts. I'm going to consider that in a new PR, and I might switch out the HLL implementation when I look into that.

jeenbroekstra commented:
Other than adding the one code comment I'm fine with this being merged.
